File: Day 8/solution.py
def part_one():
    acc = 0
    cache = []

    i = 0
    while i not in cache:
        cache.append(i)
        
        instr, value = operations[i]
        if instr == "acc":
            acc += int(value)
        elif instr == "jmp":
            i += int(value)
            continue          
        i += 1
        
    return acc


def part_two():
    for n in range(len(operations)):
        acc = 0
        cache = []
        
        i = 0
        while i >= 0 and i < len(operations):
            if i in cache:
                res = None
                break
            
            cache.append(i)
            
            instr, value = operations[i]

            if i == n:
                instr = {'jmp': 'nop', 'nop': 'jmp'}.get(instr, instr)

            if instr == 'acc':
                acc += int(value)

            elif instr == 'jmp':
                i += int(value)
                continue
            i += 1
            
            res = acc

        if res:
            return res


if __name__ == '__main__':
    fin = open("input.txt").read().split("\n")
    operations = [op.split() for op in fin]

    print("Part 1:", part_one())
    print("Part 2:", part_two())

File: Day 8/solution.rs

fn main() {
    let inp = load_input(r"C:\Users\florm\Desktop\input.txt");
    println!("Part 1: {}", part_one(&inp));
}

fn load_input(filename: &str) -> Vec<Vec<String>> {
    let operations = std::fs::read_to_string(filename)
        .expect("Couldn't open the file.");

    operations
        .lines()
        .map(|x| x.split_whitespace()
                 .map(str::to_owned)
                 .collect())
        .collect()
}

fn part_one(operations: &Vec<Vec<String>>) -> i32 {
    let mut acc = 0;
    let mut cache = vec![];

    let mut i: i32 = 0;
    while !cache.contains(&i) {
        cache.push(i);

        let instr = &operations[i as usize][0];
        let value: i32 = (&operations[i as usize][1]).parse().unwrap();
        
        if instr == "acc" {
            acc += value;
        } else if instr == "jmp" {
            i += value;
            continue;
        }
        i += 1;
    } 
    acc as i32
}